File: Data Structure/Array or Matrix/Medium/LET_no57_Insert Interval.py
class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        n = len(intervals)
        res = []
        i = j = 0
        while i < n or j == 0:
            if j > 0 or i < n and intervals[i][0] < newInterval[0]:
                self.helper(res, intervals[i][0], intervals[i][1])
                i += 1
            else:
                self.helper(res, newInterval[0], newInterval[1])
                j += 1
        return res
    
    def helper(self, res, start, end):
        if not res or start > res[-1][1]:
            res.append([start, end])
        else:
            res[-1][1] = max(res[-1][1], end)

class Solution2:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        n = len(intervals)
        cur = 0
        res = []
        
        while cur < n and intervals[cur][1] < newInterval[0]:
            res.append(intervals[cur])
            cur += 1
        
        while cur < n and intervals[cur][0] <= newInterval[1]:
            newInterval[0] = min(newInterval[0], intervals[cur][0])
            newInterval[1] = max(newInterval[1], intervals[cur][1])
            cur += 1
        
        res.append(newInterval)
        res.extend(intervals[cur:])
        
        return res

File: Data Structure/Array or Matrix/Medium/LET_no57_Insert Interval.cpp
class Solution {
public:
    vector<vector<int>> insert(vector<vector<int>>& intervals, vector<int>& newInterval) {
        int n = intervals.size(), i = 0;
        bool used = false;
        vector<vector<int>> res;

        while (i < n or !used) {
            if (used || i < n && intervals[i][0] < newInterval[0]) {
                updateRes(res, intervals[i][0], intervals[i][1]);
                i++;
            }
            else {
                updateRes(res, newInterval[0], newInterval[1]);
                used = true;
            }
        }

        return res;
    }

    void updateRes(vector<vector<int>>& res, int start, int end) {
        if (res.empty() || start > res.back()[1])
            res.push_back({start, end});
        else
            res.back()[1] = max(res.back()[1], end);
    }
};